subsys/modem/modem_cmux.c

```diff
@@ -1314,25 +1364,36 @@ int modem_cmux_disconnect(struct modem_cmux *cmux)
 
 int modem_cmux_disconnect_async(struct modem_cmux *cmux)
 {
-	__ASSERT_NO_MSG(cmux->pipe != NULL);
+	int ret = 0;
 
 	if (k_event_test(&cmux->event, MODEM_CMUX_EVENT_DISCONNECTED_BIT)) {
 		return -EALREADY;
 	}
 
-	if (k_work_delayable_is_pending(&cmux->disconnect_work) == false) {
-		k_work_schedule(&cmux->disconnect_work, K_NO_WAIT);
+	K_SPINLOCK(&cmux->work_lock) {
+		if (!cmux->attached) {
+			ret = -EPERM;
+			K_SPINLOCK_BREAK;
+		}
+
+		if (k_work_delayable_is_pending(&cmux->disconnect_work) == false) {
+			k_work_schedule(&cmux->disconnect_work, K_NO_WAIT);
+		}
 	}
 
-	return 0;
+	return ret;
 }
 
 void modem_cmux_release(struct modem_cmux *cmux)
 {
 	struct k_work_sync sync;
 
-	/* Close DLCI pipes */
-	modem_cmux_dlci_pipes_notify_closed(cmux);
+	K_SPINLOCK(&cmux->work_lock) {
```

Comment:
`attached` is used to prevent any new work from being submitted by the pipes (DLCI channels), and the modem_pipe_release ensures no new work is submitted by the bus pipe callback. The release flow is as follows:

1. Lock spinlock and clear `attached` to ensure DLCI channel pipe implementation does not submit new work
2. Release bus pipe using modem_pipe_release() which ensures receive ready/transmit idle does not submit any new work
3. Cancel all work synchronously
4. Set pipe to NULL as we are now certain the pipe is no longer in use.

We can't change the pipe until we have done step 1 2 and 3 :)

Reply:
The `modem_cmux_attach()` and `modem_cmux_release()` are only performed by the thread which owns the cmux module, so there is no scenario where `modem_cmux_attach()` is called simultaneously to `modem_cmux_release()`